File: zig/src/tools/eval_profiles.zig
const std = @import("std");
const zeroclaw = @import("zeroclaw");
const auth = zeroclaw.providers.auth;
const datetime = zeroclaw.api.datetime;

const EvalError = error{InvalidScenario};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try std.io.getStdIn().readToEndAlloc(allocator, 32 * 1024 * 1024);
    defer allocator.free(input);

    const stdout = std.io.getStdOut().writer();
    var lines = std.mem.splitScalar(u8, input, '\n');
    var index: usize = 0;
    while (lines.next()) |raw_line| {
        const line = std.mem.trim(u8, raw_line, " \t\r\n");
        if (line.len == 0) continue;
        try runOp(allocator, line, stdout, index);
        index += 1;
    }
}

fn runOp(allocator: std.mem.Allocator, line: []const u8, writer: anytype, index: usize) !void {
    var parsed = try std.json.parseFromSlice(std.json.Value, allocator, line, .{});
    defer parsed.deinit();
    const op = getString(parsed.value, "op") orelse return EvalError.InvalidScenario;

    if (std.mem.eql(u8, op, "roundtrip_oauth_profile")) {
        try roundtripOauth(allocator, parsed.value, writer, index);
    } else if (std.mem.eql(u8, op, "roundtrip_token_profile")) {
        try roundtripToken(allocator, parsed.value, writer, index);
    } else if (std.mem.eql(u8, op, "select_profile_id")) {
        try selectProfile(allocator, parsed.value, writer);
    } else if (std.mem.eql(u8, op, "migrate_legacy_enc_in_profile")) {
        try migrateLegacy(allocator, parsed.value, writer, index);
    } else if (std.mem.eql(u8, op, "schema_version_mismatch")) {
        try schemaVersionMismatch(allocator, parsed.value, writer, index);
    } else {
        return EvalError.InvalidScenario;
    }
}

fn roundtripOauth(allocator: std.mem.Allocator, value: std.json.Value, writer: anytype, index: usize) !void {
    const encrypt_secrets = getBool(value, "encrypt_secrets") orelse true;
    const now = getI64(value, "now_unix_seconds") orelse 1_700_000_000;
    const profile_value = getObject(value, "profile") orelse return EvalError.InvalidScenario;
    const state_dir = try makeTempStateDir(allocator, "profiles-oauth", index);
    defer cleanupTempStateDir(allocator, state_dir);
    if (getString(value, "key_hex")) |key_hex| try writeKeyFile(allocator, state_dir, key_hex);

    var store = try auth.AuthProfilesStore.new(allocator, state_dir, encrypt_secrets);
    defer store.deinit();
    var profile = try oauthProfileFromJson(allocator, profile_value, now);
    defer profile.deinit(allocator);
    try store.upsertProfile(&profile, true, now);
    var data = try store.load();
    defer data.deinit(allocator);
    const loaded = data.profiles.get(profile.id).?;
    try writer.writeAll("{\"op\":\"roundtrip_oauth_profile\",\"result\":{\"profile\":");
    try writeProfileJson(writer, allocator, loaded);
    try writer.writeAll("}}\n");
}

fn roundtripToken(allocator: std.mem.Allocator, value: std.json.Value, writer: anytype, index: usize) !void {
    const encrypt_secrets = getBool(value, "encrypt_secrets") orelse true;
    const now = getI64(value, "now_unix_seconds") orelse 1_700_000_000;
    const profile_value = getObject(value, "profile") orelse return EvalError.InvalidScenario;
    const state_dir = try makeTempStateDir(allocator, "profiles-token", index);
    defer cleanupTempStateDir(allocator, state_dir);
    if (getString(value, "key_hex")) |key_hex| try writeKeyFile(allocator, state_dir, key_hex);

    var store = try auth.AuthProfilesStore.new(allocator, state_dir, encrypt_secrets);
    defer store.deinit();
    var profile = try tokenProfileFromJson(allocator, profile_value, now);
    defer profile.deinit(allocator);
    try store.upsertProfile(&profile, true, now);
    var data = try store.load();
    defer data.deinit(allocator);
    const loaded = data.profiles.get(profile.id).?;
    try writer.writeAll("{\"op\":\"roundtrip_token_profile\",\"result\":{\"profile\":");
    try writeProfileJson(writer, allocator, loaded);
    try writer.writeAll("}}\n");
}

fn selectProfile(allocator: std.mem.Allocator, value: std.json.Value, writer: anytype) !void {
    var data = auth.AuthProfilesData.init(allocator, 1_700_000_000);
    defer data.deinit(allocator);
    const profiles = getArray(value, "profiles") orelse return EvalError.InvalidScenario;
    for (profiles.items) |item| {
        const provider = getString(item, "provider") orelse return EvalError.InvalidScenario;
        const name = getString(item, "name") orelse return EvalError.InvalidScenario;
        var profile = try auth.AuthProfile.newToken(allocator, provider, name, "token", 1_700_000_000);
        defer profile.deinit(allocator);
        try putProfile(allocator, &data.profiles, &profile);
    }
    if (getObject(value, "active_map")) |active_map| {
        var it = active_map.object.iterator();
        while (it.next()) |entry| {
            if (entry.value_ptr.* != .string) return EvalError.InvalidScenario;
            try putStringValue(allocator, &data.active_profiles, entry.key_ptr.*, entry.value_ptr.string);
        }
    }
    const provider = getString(value, "query_provider") orelse return EvalError.InvalidScenario;
    const override = getOptionalString(value, "override");
    const resolved = try auth.selectProfileId(allocator, &data, provider, override);
    defer if (resolved) |id| allocator.free(id);
    try writer.writeAll("{\"op\":\"select_profile_id\",\"result\":{\"resolved_id\":");
    if (resolved) |id| try std.json.stringify(id, .{}, writer) else try writer.writeAll("null");
    try writer.writeAll("}}\n");
}

fn migrateLegacy(allocator: std.mem.Allocator, value: std.json.Value, writer: anytype, index: usize) !void {
    const state_dir = try makeTempStateDir(allocator, "profiles-migrate", index);
    defer cleanupTempStateDir(allocator, state_dir);
    try writeKeyFile(allocator, state_dir, getString(value, "key_hex") orelse return EvalError.InvalidScenario);
    try writeProfilesFile(allocator, state_dir, getString(value, "file_json") orelse return EvalError.InvalidScenario);
    var store = try auth.AuthProfilesStore.new(allocator, state_dir, true);
    defer store.deinit();
    var data = try store.load();
    defer data.deinit(allocator);
    const id = getString(value, "profile_id") orelse return EvalError.InvalidScenario;
    const profile = data.profiles.get(id) orelse return EvalError.InvalidScenario;
    const access_token = profile.token_set.?.access_token;
    const raw = try std.fs.cwd().readFileAlloc(allocator, store.path(), 1024 * 1024);
    defer allocator.free(raw);
    try writer.writeAll("{\"op\":\"migrate_legacy_enc_in_profile\",\"result\":{\"access_token\":");
    try std.json.stringify(access_token, .{}, writer);
    try writer.writeAll(",\"file_now_enc2\":");
    try writer.writeAll(if (std.mem.indexOf(u8, raw, "\"access_token\": \"enc2:") != null) "true" else "false");
    try writer.writeAll("}}\n");
}

fn schemaVersionMismatch(allocator: std.mem.Allocator, value: std.json.Value, writer: anytype, index: usize) !void {
    const state_dir = try makeTempStateDir(allocator, "profiles-schema", index);
    defer cleanupTempStateDir(allocator, state_dir);
    try writeProfilesFile(allocator, state_dir, getString(value, "file_json") orelse return EvalError.InvalidScenario);
    var store = try auth.AuthProfilesStore.new(allocator, state_dir, false);
    defer store.deinit();
    const result = store.load();
    if (result) |data| {
        var owned = data;
        owned.deinit(allocator);
        return EvalError.InvalidScenario;
    } else |err| {
        try writer.writeAll("{\"op\":\"schema_version_mismatch\",\"result\":{\"error\":");
        try std.json.stringify(@errorName(err), .{}, writer);
        try writer.writeAll("}}\n");
    }
}

fn oauthProfileFromJson(allocator: std.mem.Allocator, value: std.json.Value, now: i64) !auth.AuthProfile {
    var token_set = auth.TokenSet{
        .access_token = try dupStringField(allocator, value, "access_token"),
        .refresh_token = try dupOptionalStringField(allocator, value, "refresh_token"),
        .id_token = try dupOptionalStringField(allocator, value, "id_token"),
        .expires_at_utc_seconds = getI64(value, "expires_at_unix_seconds"),
        .token_type = try dupOptionalStringField(allocator, value, "token_type"),
        .scope = try dupOptionalStringField(allocator, value, "scope"),
    };
    errdefer token_set.deinit(allocator);
    var profile = try auth.AuthProfile.newOauth(
        allocator,
        getString(value, "provider") orelse return EvalError.InvalidScenario,
        getString(value, "name") orelse return EvalError.InvalidScenario,
        token_set,
        now,
    );
    errdefer profile.deinit(allocator);
    profile.account_id = try dupOptionalStringField(allocator, value, "account_id");
    profile.workspace_id = try dupOptionalStringField(allocator, value, "workspace_id");
    try fillMetadata(allocator, &profile.metadata, value);
    return profile;
}

fn tokenProfileFromJson(allocator: std.mem.Allocator, value: std.json.Value, now: i64) !auth.AuthProfile {
    var profile = try auth.AuthProfile.newToken(
        allocator,
        getString(value, "provider") orelse return EvalError.InvalidScenario,
        getString(value, "name") orelse return EvalError.InvalidScenario,
        getString(value, "token") orelse return EvalError.InvalidScenario,
        now,
    );
    errdefer profile.deinit(allocator);
    profile.account_id = try dupOptionalStringField(allocator, value, "account_id");
    profile.workspace_id = try dupOptionalStringField(allocator, value, "workspace_id");
    try fillMetadata(allocator, &profile.metadata, value);
    return profile;
}

fn writeProfileJson(writer: anytype, allocator: std.mem.Allocator, profile: auth.AuthProfile) !void {
    try writer.writeByte('{');
    var wrote = false;
    try writeStringField(writer, &wrote, "id", profile.id);
    try writeStringField(writer, &wrote, "provider", profile.provider);
    try writeStringField(writer, &wrote, "profile_name", profile.profile_name);
    try writeStringField(writer, &wrote, "kind", profile.kind.asString());
    try writeOptionalStringField(writer, &wrote, "account_id", profile.account_id);
    try writeOptionalStringField(writer, &wrote, "workspace_id", profile.workspace_id);
    try writeComma(writer, &wrote);
    try writer.writeAll("\"token_set\":");
    if (profile.token_set) |token_set| try writeTokenSetJson(writer, allocator, token_set) else try writer.writeAll("null");
    try writeOptionalStringField(writer, &wrote, "token", profile.token);
    try writeComma(writer, &wrote);
    try writer.writeAll("\"metadata\":");
    try writeStringMapJson(writer, allocator, profile.metadata);
    const created_at = try datetime.formatRfc3339(allocator, profile.created_at_unix_seconds);
    defer allocator.free(created_at);
    const updated_at = try datetime.formatRfc3339(allocator, profile.updated_at_unix_seconds);
    defer allocator.free(updated_at);
    try writeStringField(writer, &wrote, "created_at", created_at);
    try writeStringField(writer, &wrote, "updated_at", updated_at);
    try writer.writeByte('}');
}

fn writeTokenSetJson(writer: anytype, allocator: std.mem.Allocator, token_set: auth.TokenSet) !void {
    _ = allocator;
    try writer.writeByte('{');
    var wrote = false;
    try writeStringField(writer, &wrote, "access_token", token_set.access_token);
    try writeOptionalStringField(writer, &wrote, "refresh_token", token_set.refresh_token);
    try writeOptionalStringField(writer, &wrote, "id_token", token_set.id_token);
    try writeComma(writer, &wrote);
    try writer.writeAll("\"expires_at\":");
    if (token_set.expires_at_utc_seconds) |expires_at| {
        const formatted = try datetime.formatRfc3339(std.heap.page_allocator, expires_at);
        defer std.heap.page_allocator.free(formatted);
        try std.json.stringify(formatted, .{}, writer);
    } else {
        try writer.writeAll("null");
    }
    try writeOptionalStringField(writer, &wrote, "token_type", token_set.token_type);
    try writeOptionalStringField(writer, &wrote, "scope", token_set.scope);
    try writer.writeByte('}');
}

fn writeStringMapJson(writer: anytype, allocator: std.mem.Allocator, map: std.StringHashMap([]u8)) !void {
    const keys = try sortedKeys(allocator, map);
    defer allocator.free(keys);
    try writer.writeByte('{');
    for (keys, 0..) |key, i| {
        if (i != 0) try writer.writeByte(',');
        try std.json.stringify(key, .{}, writer);
        try writer.writeByte(':');
        try std.json.stringify(map.get(key).?, .{}, writer);
    }
    try writer.writeByte('}');
}

fn writeStringField(writer: anytype, wrote: *bool, key: []const u8, value: []const u8) !void {
    try writeComma(writer, wrote);
    try std.json.stringify(key, .{}, writer);
    try writer.writeByte(':');
    try std.json.stringify(value, .{}, writer);
}

fn writeOptionalStringField(writer: anytype, wrote: *bool, key: []const u8, value: ?[]const u8) !void {
    try writeComma(writer, wrote);
    try std.json.stringify(key, .{}, writer);
    try writer.writeByte(':');
    if (value) |inner| try std.json.stringify(inner, .{}, writer) else try writer.writeAll("null");
}

fn writeComma(writer: anytype, wrote: *bool) !void {
    if (wrote.*) try writer.writeByte(',');
    wrote.* = true;
}

fn fillMetadata(allocator: std.mem.Allocator, map: *std.StringHashMap([]u8), value: std.json.Value) !void {
    const metadata = getObject(value, "metadata") orelse return;
    var it = metadata.object.iterator();
    while (it.next()) |entry| {
        if (entry.value_ptr.* != .string) return EvalError.InvalidScenario;
        try putStringValue(allocator, map, entry.key_ptr.*, entry.value_ptr.string);
    }
}

fn putProfile(allocator: std.mem.Allocator, map: *std.StringHashMap(auth.AuthProfile), profile: *const auth.AuthProfile) !void {
    const clone = try profile.clone(allocator);
    errdefer {
        var tmp = clone;
        tmp.deinit(allocator);
    }
    const gop = try map.getOrPut(profile.id);
    if (!gop.found_existing) gop.key_ptr.* = try allocator.dupe(u8, profile.id);
    gop.value_ptr.* = clone;
}

fn putStringValue(allocator: std.mem.Allocator, map: *std.StringHashMap([]u8), key: []const u8, value: []const u8) !void {
    const value_owned = try allocator.dupe(u8, value);
    errdefer allocator.free(value_owned);
    const gop = try map.getOrPut(key);
    if (gop.found_existing) {
        allocator.free(gop.value_ptr.*);
    } else {
        gop.key_ptr.* = try allocator.dupe(u8, key);
    }
    gop.value_ptr.* = value_owned;
}

fn sortedKeys(allocator: std.mem.Allocator, map: std.StringHashMap([]u8)) ![][]const u8 {
    var keys = try allocator.alloc([]const u8, map.count());
    var i: usize = 0;
    var it = map.iterator();
    while (it.next()) |entry| {
        keys[i] = entry.key_ptr.*;
        i += 1;
    }
    std.mem.sort([]const u8, keys, {}, lessThanBytes);
    return keys;
}

fn lessThanBytes(_: void, a: []const u8, b: []const u8) bool {
    return std.mem.order(u8, a, b) == .lt;
}

fn makeTempStateDir(allocator: std.mem.Allocator, prefix: []const u8, index: usize) ![]u8 {
    const dir = try std.fmt.allocPrint(
        allocator,
        "/tmp/zeroclaw-eval-{s}-zig-{d}-{d}",
        .{ prefix, if (@hasDecl(std.c, "getpid")) std.c.getpid() else 0, index },
    );
    errdefer allocator.free(dir);
    std.fs.cwd().deleteTree(dir) catch {};
    try std.fs.cwd().makePath(dir);
    return dir;
}

fn cleanupTempStateDir(allocator: std.mem.Allocator, path: []u8) void {
    std.fs.cwd().deleteTree(path) catch {};
    allocator.free(path);
}

fn writeKeyFile(allocator: std.mem.Allocator, state_dir: []const u8, key_hex: []const u8) !void {
    const path = try std.fs.path.join(allocator, &.{ state_dir, ".secret_key" });
    defer allocator.free(path);
    var file = try std.fs.cwd().createFile(path, .{ .truncate = true, .mode = 0o600 });
    defer file.close();
    try file.writeAll(key_hex);
    try file.chmod(0o600);
}

fn writeProfilesFile(allocator: std.mem.Allocator, state_dir: []const u8, contents: []const u8) !void {
    const path = try std.fs.path.join(allocator, &.{ state_dir, "auth-profiles.json" });
    defer allocator.free(path);
    var file = try std.fs.cwd().createFile(path, .{ .truncate = true });
    defer file.close();
    try file.writeAll(contents);
}

fn getString(value: std.json.Value, key: []const u8) ?[]const u8 {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner != .string) return null;
    return inner.string;
}

fn getOptionalString(value: std.json.Value, key: []const u8) ?[]const u8 {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner == .null) return null;
    if (inner != .string) return null;
    return inner.string;
}

fn dupStringField(allocator: std.mem.Allocator, value: std.json.Value, key: []const u8) ![]u8 {
    return allocator.dupe(u8, getString(value, key) orelse return EvalError.InvalidScenario);
}

fn dupOptionalStringField(allocator: std.mem.Allocator, value: std.json.Value, key: []const u8) !?[]u8 {
    if (getOptionalString(value, key)) |inner| return try allocator.dupe(u8, inner);
    return null;
}

fn getBool(value: std.json.Value, key: []const u8) ?bool {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner != .bool) return null;
    return inner.bool;
}

fn getI64(value: std.json.Value, key: []const u8) ?i64 {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner != .integer) return null;
    return inner.integer;
}

fn getObject(value: std.json.Value, key: []const u8) ?std.json.Value {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner != .object) return null;
    return inner;
}

fn getArray(value: std.json.Value, key: []const u8) ?std.json.Array {
    if (value != .object) return null;
    const inner = value.object.get(key) orelse return null;
    if (inner != .array) return null;
    return inner.array;
}

File: eval-tools/src/bin/eval-profiles.rs
//! eval-profiles — AuthProfilesStore/AuthService helper parity runner.

use std::collections::BTreeMap;
use std::fs;
use std::io::{self, Read, Write};
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use serde_json::{Map, Value};
use zeroclaw_providers::auth::profiles::{AuthProfile, AuthProfilesData, AuthProfilesStore, TokenSet};
use zeroclaw_providers::auth::select_profile_id;

#[tokio::main(flavor = "current_thread")]
async fn main() -> Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let mut output = Vec::new();

    for (index, raw_line) in input.lines().enumerate() {
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }
        let value: Value = serde_json::from_str(line).context("invalid scenario JSON")?;
        let op = required_string(&value, "op")?;
        run_op(&mut output, index, op, &value).await?;
    }

    io::stdout().lock().write_all(&output)?;
    Ok(())
}

async fn run_op(output: &mut Vec<u8>, index: usize, op: &str, value: &Value) -> Result<()> {
    match op {
        "roundtrip_oauth_profile" => roundtrip_oauth(output, index, op, value).await,
        "roundtrip_token_profile" => roundtrip_token(output, index, op, value).await,
        "select_profile_id" => select_profile(output, op, value),
        "migrate_legacy_enc_in_profile" => migrate_legacy(output, index, op, value).await,
        "schema_version_mismatch" => schema_version_mismatch(output, index, op, value).await,
        _ => anyhow::bail!("unknown profiles op: {op}"),
    }
}

async fn roundtrip_oauth(output: &mut Vec<u8>, index: usize, op: &str, value: &Value) -> Result<()> {
    let encrypt_secrets = value.get("encrypt_secrets").and_then(Value::as_bool).unwrap_or(true);
    let state_dir = make_temp_state_dir(index, "profiles-oauth-rust")?;
    if let Some(key_hex) = value.get("key_hex").and_then(Value::as_str) {
        write_key_file(&state_dir, key_hex)?;
    }
    let store = AuthProfilesStore::new(&state_dir, encrypt_secrets);
    let mut profile = oauth_profile_from_value(required_object(value, "profile")?)?;
    store.upsert_profile(profile.clone(), true).await?;
    let data = store.load().await?;
    profile = data.profiles.get(&profile.id).context("profile missing")?.clone();
    write_result(output, op, serde_json::json!({ "profile": profile_json(&profile) }));
    let _ = fs::remove_dir_all(state_dir);
    Ok(())
}

async fn roundtrip_token(output: &mut Vec<u8>, index: usize, op: &str, value: &Value) -> Result<()> {
    let encrypt_secrets = value.get("encrypt_secrets").and_then(Value::as_bool).unwrap_or(true);
    let state_dir = make_temp_state_dir(index, "profiles-token-rust")?;
    if let Some(key_hex) = value.get("key_hex").and_then(Value::as_str) {
        write_key_file(&state_dir, key_hex)?;
    }
    let store = AuthProfilesStore::new(&state_dir, encrypt_secrets);
    let mut profile = token_profile_from_value(required_object(value, "profile")?)?;
    store.upsert_profile(profile.clone(), true).await?;
    let data = store.load().await?;
    profile = data.profiles.get(&profile.id).context("profile missing")?.clone();
    write_result(output, op, serde_json::json!({ "profile": profile_json(&profile) }));
    let _ = fs::remove_dir_all(state_dir);
    Ok(())
}

fn select_profile(output: &mut Vec<u8>, op: &str, value: &Value) -> Result<()> {
    let mut data = AuthProfilesData::default();
    for item in required_array(value, "profiles")? {
        let provider = required_string(item, "provider")?;
        let name = required_string(item, "name")?;
        let profile = AuthProfile::new_token(provider, name, "token".to_string());
        data.profiles.insert(profile.id.clone(), profile);
    }
    if let Some(active) = value.get("active_map").and_then(Value::as_object) {
        for (provider, id) in active {
            data.active_profiles
                .insert(provider.clone(), id.as_str().context("active id must be string")?.to_string());
        }
    }
    let provider = required_string(value, "query_provider")?;
    let profile_override = optional_string(value, "override")?;
    let resolved = select_profile_id(&data, provider, profile_override.as_deref());
    write_result(output, op, serde_json::json!({ "resolved_id": resolved }));
    Ok(())
}

async fn migrate_legacy(output: &mut Vec<u8>, index: usize, op: &str, value: &Value) -> Result<()> {
    let state_dir = make_temp_state_dir(index, "profiles-migrate-rust")?;
    write_key_file(&state_dir, required_string(value, "key_hex")?)?;
    write_profiles_file(&state_dir, required_string(value, "file_json")?)?;
    let store = AuthProfilesStore::new(&state_dir, true);
    let data = store.load().await?;
    let profile = data
        .profiles
        .get(required_string(value, "profile_id")?)
        .context("profile missing")?;
    let access_token = profile
        .token_set
        .as_ref()
        .context("token set missing")?
        .access_token
        .clone();
    let raw = fs::read_to_string(store.path())?;
    write_result(
        output,
        op,
        serde_json::json!({
            "access_token": access_token,
            "file_now_enc2": raw.contains("\"access_token\": \"enc2:"),
        }),
    );
    let _ = fs::remove_dir_all(state_dir);
    Ok(())
}

async fn schema_version_mismatch(output: &mut Vec<u8>, index: usize, op: &str, value: &Value) -> Result<()> {
    let state_dir = make_temp_state_dir(index, "profiles-schema-rust")?;
    write_profiles_file(&state_dir, required_string(value, "file_json")?)?;
    let store = AuthProfilesStore::new(&state_dir, false);
    let result = match store.load().await {
        Ok(_) => serde_json::json!({ "error": "None" }),
        Err(err) if err.to_string().contains("Unsupported auth profile schema version") => {
            serde_json::json!({ "error": "UnsupportedSchemaVersion" })
        }
        Err(err) => serde_json::json!({ "error": err.to_string() }),
    };
    write_result(output, op, result);
    let _ = fs::remove_dir_all(state_dir);
    Ok(())
}

fn oauth_profile_from_value(value: &Value) -> Result<AuthProfile> {
    let token_set = TokenSet {
        access_token: required_string(value, "access_token")?.to_string(),
        refresh_token: optional_string(value, "refresh_token")?,
        id_token: optional_string(value, "id_token")?,
        expires_at: None,
        token_type: optional_string(value, "token_type")?,
        scope: optional_string(value, "scope")?,
    };
    let mut profile = AuthProfile::new_oauth(
        required_string(value, "provider")?,
        required_string(value, "name")?,
        token_set,
    );
    profile.account_id = optional_string(value, "account_id")?;
    profile.workspace_id = optional_string(value, "workspace_id")?;
    profile.metadata = metadata_from_value(value)?;
    Ok(profile)
}

fn token_profile_from_value(value: &Value) -> Result<AuthProfile> {
    let mut profile = AuthProfile::new_token(
        required_string(value, "provider")?,
        required_string(value, "name")?,
        required_string(value, "token")?.to_string(),
    );
    profile.account_id = optional_string(value, "account_id")?;
    profile.workspace_id = optional_string(value, "workspace_id")?;
    profile.metadata = metadata_from_value(value)?;
    Ok(profile)
}

fn profile_json(profile: &AuthProfile) -> Value {
    let mut result = Map::new();
    result.insert("id".to_string(), serde_json::json!(profile.id));
    result.insert("provider".to_string(), serde_json::json!(profile.provider));
    result.insert("profile_name".to_string(), serde_json::json!(profile.profile_name));
    result.insert("kind".to_string(), serde_json::json!(match profile.kind {
        zeroclaw_providers::auth::profiles::AuthProfileKind::OAuth => "oauth",
        zeroclaw_providers::auth::profiles::AuthProfileKind::Token => "token",
    }));
    result.insert("account_id".to_string(), serde_json::to_value(&profile.account_id).unwrap());
    result.insert("workspace_id".to_string(), serde_json::to_value(&profile.workspace_id).unwrap());
    result.insert(
        "token_set".to_string(),
        profile
            .token_set
            .as_ref()
            .map(token_set_json)
            .unwrap_or(Value::Null),
    );
    result.insert("token".to_string(), serde_json::to_value(&profile.token).unwrap());
    result.insert("metadata".to_string(), serde_json::to_value(&profile.metadata).unwrap());
    result.insert("created_at".to_string(), serde_json::json!(profile.created_at.to_rfc3339()));
    result.insert("updated_at".to_string(), serde_json::json!(profile.updated_at.to_rfc3339()));
    Value::Object(result)
}

fn token_set_json(token_set: &TokenSet) -> Value {
    serde_json::json!({
        "access_token": token_set.access_token,
        "refresh_token": token_set.refresh_token,
        "id_token": token_set.id_token,
        "expires_at": token_set.expires_at.map(|dt| dt.to_rfc3339()),
        "token_type": token_set.token_type,
        "scope": token_set.scope,
    })
}

fn metadata_from_value(value: &Value) -> Result<BTreeMap<String, String>> {
    let mut map = BTreeMap::new();
    if let Some(metadata) = value.get("metadata").and_then(Value::as_object) {
        for (key, value) in metadata {
            map.insert(key.clone(), value.as_str().context("metadata value must be string")?.to_string());
        }
    }
    Ok(map)
}

fn make_temp_state_dir(index: usize, prefix: &str) -> Result<PathBuf> {
    let path = std::env::temp_dir().join(format!("{prefix}-{}-{index}", std::process::id()));
    let _ = fs::remove_dir_all(&path);
    fs::create_dir_all(&path)?;
    Ok(path)
}

fn write_key_file(state_dir: &Path, key_hex: &str) -> Result<()> {
    fs::write(state_dir.join(".secret_key"), key_hex)?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(state_dir.join(".secret_key"), fs::Permissions::from_mode(0o600))?;
    }
    Ok(())
}

fn write_profiles_file(state_dir: &Path, contents: &str) -> Result<()> {
    fs::create_dir_all(state_dir)?;
    fs::write(state_dir.join("auth-profiles.json"), contents)?;
    Ok(())
}

fn required_string<'a>(value: &'a Value, key: &str) -> Result<&'a str> {
    value
        .get(key)
        .and_then(Value::as_str)
        .with_context(|| format!("{key} missing"))
}

fn optional_string(value: &Value, key: &str) -> Result<Option<String>> {
    match value.get(key) {
        None | Some(Value::Null) => Ok(None),
        Some(Value::String(inner)) => Ok(Some(inner.clone())),
        Some(_) => anyhow::bail!("{key} must be string or null"),
    }
}

fn required_object<'a>(value: &'a Value, key: &str) -> Result<&'a Value> {
    let inner = value.get(key).with_context(|| format!("{key} missing"))?;
    if !inner.is_object() {
        anyhow::bail!("{key} must be object");
    }
    Ok(inner)
}

fn required_array<'a>(value: &'a Value, key: &str) -> Result<&'a Vec<Value>> {
    value
        .get(key)
        .and_then(Value::as_array)
        .with_context(|| format!("{key} missing"))
}

fn write_result(output: &mut Vec<u8>, op: &str, result: Value) {
    let value = serde_json::json!({ "op": op, "result": result });
    serde_json::to_writer(&mut *output, &value).expect("serialize eval result");
    output.push(b'\n');
}
